### PR TITLE
[OTE-456] Make FNS compatible with OE - send all block processing events in `Precommit`

### PR DESCRIPTION
**Pull request by teddyding**
### Changelist
- Store order fill stream events in a transient store during block processing
- Do the following in `Precommit`, after consensus agrees on a block. 
   - Send onchain order fills
   - Send onchain subacount updates 
   - Send OrderbookUpdates to synchronize optimistic fills in our local ops queue. 

### Test Plan
(TODO) FNS integration testing on dev

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
  - Added support for on-chain stream events with new methods for managing and retrieving these events.
  - Implemented the `Precommit` function for the clob module to handle ABCI Precommit logic.
  - Introduced `StagedFinalizeBlockEvent` and `StagedFinalizeBlockEventSDKType` interfaces to manage events during the block finalization process.

- **Bug Fixes**
  - Adjusted initialization parameters across multiple components to include additional arguments for enhanced functionality.

- **Refactor**
  - Replaced obsolete gRPC stream orderbook update logic with streamlined on-chain stream event management.

### CODE REVIEW

File: protocol/x/clob/keeper/grpc_stream_orderbook.go [hunk 1]
```diff
@@ -18,3 +20,54 @@ func (k Keeper) StreamOrderbookUpdates(
 	<-ctx.Done()
 	return nil
 }
+
+// revertOptimisticFills sends OrderUpdates that effectively reverts the optimistic fills.
+func (k Keeper) revertOptimisticFills(ctx sdk.Context) {
```
Comment (teddyding) [inline protocol/x/clob/keeper/grpc_stream_orderbook.go:25]: Moved from [ProcessProposerOps](https://github.com/dydxprotocol/v4-chain/blob/e09241b99990611e89eca40737ff1df1ab2416f2/protocol/x/clob/keeper/process_operations.go#L62-L78).

File: protocol/x/clob/keeper/grpc_stream_orderbook.go [hunk 2]
```diff
@@ -18,3 +20,54 @@ func (k Keeper) StreamOrderbookUpdates(
...
+	// Send absolute fill amounts from local to the grpc stream.
+	// This must be sent out to account for checkState being discarded and deliverState being used.
+	localValidatorOperationsQueue, _ := k.MemClob.GetOperationsToReplay(ctx)
+	orderIdsFromLocalOpsQueue := fetchOrdersInvolvedInOpQueue(
```
Comment (teddyding) [inline protocol/x/clob/keeper/grpc_stream_orderbook.go:29]: No longer fetching orders from proposer ops queue, which from offline discussion we agree isn't really needed

File: protocol/indexer/indexer_manager/event_manager.go [hunk 3]
```diff
@@ -3,12 +3,17 @@ package indexer_manager
...
 	storetypes "cosmossdk.io/store/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	ante_types "github.com/dydxprotocol/v4-chain/protocol/app/ante/types"
 	"github.com/dydxprotocol/v4-chain/protocol/indexer/msgsender"
 )
 
 type IndexerEventManager interface {
 	Enabled() bool
 	AddTxnEvent(ctx sdk.Context, subType string, version uint32, dataByes []byte)
+	AddOnchainStreamEvent(
```
Comment (jonfung-dydx) [inline protocol/indexer/indexer_manager/event_manager.go:13]: can we store this on the grpc streaming manager instead?

File: protocol/indexer/indexer_manager/event_manager.go [hunk 4]
```diff
@@ -28,17 +34,21 @@ var _ IndexerEventManager = (*indexerEventManagerImpl)(nil)
 type indexerEventManagerImpl struct {
 	indexerMessageSender           msgsender.IndexerMessageSender
 	indexerEventsTransientStoreKey storetypes.StoreKey
+	// stores the onchain events for full node streaming.
+	onchainStreamTransientStoreKey storetypes.StoreKey
```
Comment (jonfung-dydx) [inline protocol/indexer/indexer_manager/event_manager.go:38]: nit: grpcStreamOnchainTransientStoreKey

File: protocol/indexer/indexer_manager/event_manager.go [hunk 5]
```diff
@@ -64,6 +74,20 @@ func (i *indexerEventManagerImpl) SendOnchainData(block *IndexerTendermintBlock)
...
+func (i *indexerEventManagerImpl) AddOnchainStreamEvent(
+	ctx sdk.Context,
+	dataBytes []byte,
+) {
+	event := IndexerTendermintEventWrapper{
+		Event: &IndexerTendermintEvent{
+			DataBytes: dataBytes,
+		},
+	}
+	addEvent(ctx, event, i.onchainStreamTransientStoreKey)
```
Comment (jonfung-dydx) [inline protocol/indexer/indexer_manager/event_manager.go:88]: The code here https://github.com/dydxprotocol/v4-chain/blob/e09241b99990611e89eca40737ff1df1ab2416f2/protocol/indexer/indexer_manager/events.go#L118
increments indexer `IndexerEventsCountKey`. We shouldn't do that for transient events right?

File: protocol/x/clob/keeper/grpc_stream_orderbook.go [hunk 6]
```diff
@@ -18,3 +20,54 @@ func (k Keeper) StreamOrderbookUpdates(
...
+	for orderId := range orderIdsFromLocalOpsQueue {
+		orderbookUpdate := k.MemClob.GetOrderbookUpdatesForOrderUpdate(ctx, orderId)
+		allUpdates.Append(orderbookUpdate)
+	}
+	k.SendOrderbookUpdates(ctx, allUpdates)
+}
+
+// streamOnchainFills sends updates resulting from onchain fills.
+func (k Keeper) streamOnchainFills(ctx sdk.Context) {
+	// Get onchain stream events stored in transient store.
```
Comment (jonfung-dydx) [inline protocol/x/clob/keeper/grpc_stream_orderbook.go:43]: nit: add a comment about how the optimistic fills are discarded from transient store if OE fails

**Pull request closed without merging**